### PR TITLE
feat: add fixed-size unsigned integer for ECLAIR

### PR DESCRIPTION
**Pull request by tsunrise**
Add `AssertWithinRange` trait and `U128` type for ECLAIR.

Closes: #141 

---

Before we can merge this PR, please make sure that all the following items have been checked off:

- [x] Linked to an issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Added **one** line describing your change in [`CHANGELOG.md`](https://github.com/manta-network/manta-rs/blob/main/CHANGELOG.md) and added the appropriate `changelog` label to the PR.
- [x] Re-reviewed `Files changed` in the GitHub PR explorer.
- [x] Checked that changes and commits conform to the standards outlined in [`CONTRIBUTING.md`](https://github.com/manta-network/manta-rs/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: manta-crypto/src/eclair/uint.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+//! Unsigned Integers Implementation
+use crate::{
+    constraint::{Add, Sub},
+    eclair::{
+        bool::AssertWithinRange,
+        ops::{AddAssign, SubAssign},
+    },
+};
+
+/// 128-Bit Unsigned Integer.
```
Comment (BoyuanFeng) [inline manta-crypto/src/eclair/uint.rs:26]: `128-Bit Unsigned Integer`

File: manta-crypto/src/eclair/uint.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+where
+    COM: AssertWithinRange<T, 128>,
+    T: Sub<T, COM, Output = T>,
+{
+    type Output = Self;
+
+    // TODO: Should we check self.0 > rhs.0 before subtraction?
+    // TODO: Should we check self > rhs instead of `AssertWithinRange`?
+    #[inline]
+    fn sub(self, rhs: Self, compiler: &mut COM) -> Self::Output {
```
Comment (BoyuanFeng) [inline manta-crypto/src/eclair/uint.rs:76]: Should we have a trait for `Saturated Subtraction`?

File: manta-pay/src/crypto/constraint/arkworks/mod.rs [hunk 3]
```diff
@@ -43,6 +44,7 @@ use manta_util::serde::{Deserialize, Serialize, Serializer};
 
 pub use ark_r1cs::SynthesisError;
 pub use ark_r1cs_std::{bits::boolean::Boolean, fields::fp::FpVar};
+use manta_crypto::eclair::bool::AssertWithinRange;
```
Comment (BoyuanFeng) [inline manta-pay/src/crypto/constraint/arkworks/mod.rs:47]: Should be moved above?

**Review by bhgomes — CHANGES_REQUESTED** [review]

File: manta-crypto/src/eclair/uint.rs (hunk 2 above)
Comment (bhgomes) [reply manta-crypto/src/eclair/uint.rs:76]: Yeah, let's forget subtraction for now: move to an issue.

Comment (bhgomes) [reply manta-crypto/src/eclair/uint.rs:76]: @tsunrise can you delete the subtraction implementation and move it to an issue?

File: manta-pay/src/crypto/constraint/arkworks/mod.rs [hunk 4]
```diff
@@ -334,6 +336,35 @@ where
     // TODO: Implement these optimizations.
 }
 
+impl<F, const BITS: usize> AssertWithinRange<FpVar<F>, BITS> for R1CS<F>
+where
+    F: PrimeField,
+{
+    #[inline]
+    fn assert_within_range(&mut self, value: &FpVar<F>) {
```
Comment (bhgomes) [inline manta-pay/src/crypto/constraint/arkworks/mod.rs:344]: Add `BITS` validity assertion at the top (using Rust `assert!`).

**Review by bhgomes — CHANGES_REQUESTED** [review]
@BoyuanFeng and I are going to rewrite some of this (and add some bit-decomposition optimizations).

**Comment by tsunrise** [comment]
> @BoyuanFeng and I are going to rewrite some of this (and add some bit-decomposition optimizations).

Sounds good, thanks!

**Review by BoyuanFeng — APPROVED** [review]

**Review by BoyuanFeng — CHANGES_REQUESTED** [review]
Need to add a test:
1) Anything lower than 2^128 needs to pass
2) Anything larger than 2^128 needs to fail
3) Edge case: 2^128 should fail; 2^128-1 should pass;

**Review by bhgomes — CHANGES_REQUESTED** [review]
Some comments on the tests. Also, @BoyuanFeng don't forget the following:

- Add generic implementation for `UnsignedInteger` from yesterday's discussion
- Lets also add an unwrapping method for `UnsignedInteger`:
    ```rust
        pub fn into_inner(self) -> T {
            self.0
        }
    ```

File: manta-pay/src/crypto/constraint/arkworks/mod.rs [hunk 5]
```diff
@@ -548,3 +569,56 @@ where
...
+        }
+        // anything greater or equal to 2^128 needs to fail
+        let fp128 = Fp(field_new!(Fr, "2").pow(&[128u64]));
+        for _ in 0..NUM_TESTS {
+            let mut value = rng.gen::<_, Fp<Fr>>();
+            // It's quite unlikely that the value is less than 2^128. We resample it until we get one that is.
+            while value < fp128 {
+                value = rng.gen::<_, Fp<Fr>>();
+            }
+            assert_within_range::<BITS>(value, false);
```
[the remark this answers is not in the document]
Comment (bhgomes) [reply manta-pay/src/crypto/constraint/arkworks/mod.rs:617]: Also, lets change it to `<=` just to be safe (since we do the `= 2^128` check later anyway).

**Review by BoyuanFeng — CHANGES_REQUESTED** [review]

**Review by bhgomes — DISMISSED** [review]

**Review by BoyuanFeng — APPROVED** [review]

**Review by bhgomes — APPROVED** [review]

**Review by BoyuanFeng — APPROVED** [review]
